spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/jdbc/DataSourceHealthContributorAutoConfiguration.java

```diff
@@ -108,14 +122,29 @@ private String getValidationQuery(DataSource source) {
 	}
 
 	/**
-	 * {@link HealthIndicator} used for {@link AbstractRoutingDataSource} beans where we
-	 * can't actually query for the status.
+	 * {@link CompositeHealthContributor} used for {@link AbstractRoutingDataSource} beans
+	 * where the overall health is composed of a {@link DataSourceHealthIndicator} for
+	 * each routed datasource.
 	 */
-	static class RoutingDataSourceHealthIndicator extends AbstractHealthIndicator {
+	static class RoutingDataSourceHealthIndicator implements CompositeHealthContributor {
+
+		private CompositeHealthContributor delegate;
+
+		RoutingDataSourceHealthIndicator(AbstractRoutingDataSource routingDataSource,
+				Function<DataSource, HealthContributor> indicatorFunction) {
+			Map<String, DataSource> routedDataSources = routingDataSource.getResolvedDataSources().entrySet().stream()
+					.collect(toMap((e) -> e.getKey().toString(), Map.Entry::getValue));
+			delegate = CompositeHealthContributor.fromMap(routedDataSources, indicatorFunction);
```

Comment:
We could have simply returned this `CompositeHealthContributor.fromMap(routedDataSources, indicatorFunction)` from [here](https://github.com/spring-projects/spring-boot/pull/25708/files#diff-3c55a3e9dc27c7e0efe132a5e0251634b7b86dfe54f65143f3233f5219f01601R114) and not created a 1st class type for it. However, its helpful for testing as well as for later extensions to the indicator if needed.